### DISCUSSION TITLE
How to configure Gradle, Github Package Registry Maven

### DISCUSSION THREAD

**Discussion by trevjonez** [community/community#26483] [Packages]
<p>I was just playing with this and thought I would share what I have found so far.<br>
Configuration is fairly straight forward when using the <code>maven-publish</code> plugin.<br>
<a href="https://docs.gradle.org/current/userguide/publishing_maven.html" rel="nofollow noopener">https://docs.gradle.org/current/userguide/publishing_maven.html</a></p>
<pre><code>apply plugin: "maven-publish"

publishing {
  repositories {
    maven {
      name = "Github"
      url = uri("https://maven.pkg.github.com/owner/repo") //TODO update owner/repo
      credentials {
        //findProperty will allow these properties to be absent and still run the build, just won't publish
        username = findProperty("GITHUB_USERNAME")
        password = findProperty("GITHUB_API_TOKEN")
      }
    }
  }
  publications {
    register("jar", MavenPublication) {
      from(components["java"])
      pom {
        url.set("https://github.com/owner/repo.git") //TODO update owner/repo
      }
    }
  }
}
</code></pre>
<p>I am unable to upload classified artifacts, IE sources jar.</p>
<pre><code>Could not PUT 'https://maven.pkg.github.com/owner/repo/com/owner/repo/x.y.z/ArtifactId-x.y.z-sources.jar'. Received status code 403 from server: Forbidden
</code></pre>
<p> I sent in a support request on the classified artifacts being rejected, as well as a request to add gradle to the documentation. </p>
<p>My hope is that they will also have decent support for gradle projects producing <code>.module</code> metadata by the time it reaches general release as well.<br>
<a href="https://github.com/gradle/gradle/blob/f6a98158e75a636245f70d46604fcab3152361e8/subprojects/docs/src/docs/design/gradle-module-metadata-1.0-specification.md" rel="nofollow noopener">https://github.com/gradle/gradle/blob/f6a98158e75a636245f70d46604fcab3152361e8/subprojects/docs/src/docs/design/gradle-module-metadata-1.0-specification.md</a></p>

**Comment by MarvinSchramm** [comment]
<p>Hej man any news on that? </p>
<p>I have the same issue that my sources.jar gets rejected</p>

**Comment by trevjonez** [comment]
<p><a class="mention" href="/marvinschramm">@marvinschramm</a> They responded asking for reproduction steps which I provided and have not heard anything since then.</p>
<p>additionally they don’t handle the maven-metadata.xml correctly which results in 400 errors when gradle tries to read what is there before publishing a new snapshot.</p>

**Comment by MarvinSchramm** [comment]
<p>Yea i’ve noticed that.</p>
<p>Now we have 1 half published package on our repository which we cant delete…</p>
<p>I think we’ll wait until everything is out of beta.</p>
<p>Thank you!</p>

**Comment by mgodave** [comment]
<p>I have the same issues, I’ve had nothing but issues, this looks like such a useful thing but it seems funamentally broken at the moment.</p>

**Comment by njpsanghani** [comment]
<p>Same Here! <a class="mention" href="/marvinschramm">@marvinschramm</a> did it work for u???</p>

**Comment by liach** [comment]
<p>Same problems from me. I am able to publish the pom files with package registry, but beyond that, everything is rejected. People said that I might need pom properties in my jar, but that doesn’t help in any way.</p>
<p>Addiiton of pom properties:</p>
<p><a href="https://github.com/liachmodded/carttemplate/blob/161758cf93b6317909dcbaa411a06303503acfee/build.gradle#L265" rel="nofollow noopener">https://github.com/liachmodded/carttemplate/blob/161758cf93b6317909dcbaa411a06303503acfee/build.gradle#L265</a></p>
<p>The packages with pom only:</p>
<p><a href="https://github.com/liachmodded/carttemplate/packages/21483" rel="nofollow noopener">https://github.com/liachmodded/carttemplate/packages/21483</a></p>

**Comment by BartvHelvert** [comment]
<p>I just ran into the same issue. For me it did upload the build .jar and the .pom file but it rejected the sources and javadoc. The publication can be found here: <a href="https://github.com/guthix/Jagex-Store-5/packages/32130" rel="nofollow noopener">https://github.com/guthix/Jagex-Store-5/packages/32130</a>. </p>

**Comment by ajcann** [comment]
<p><a class="mention" href="/trevjonez">@trevjonez</a> Running into the same issue as well. Support ticket in but nothing yet. Any updates on your side?</p>

**Comment by trevjonez** [comment]
<p>I got an initial response that was “can you give a reproduction example” for the classified artifact issue, which I gave them immediately. Radio silence since then.</p>
<p>here is my reproduction: <a href="https://github.com/trevjonez/github-registry-repro" rel="nofollow noopener">https://github.com/trevjonez/github-registry-repro</a></p>
<p>as well as a gradle scan of the issue happening: <a href="https://scans.gradle.com/s/eqnfwqbyopiyu/" rel="nofollow noopener">https://scans.gradle.com/s/eqnfwqbyopiyu/</a></p>
<p>while doing the reproduction I also found that republishing snapshots fail when querying the maven metadata which is shown in this scan: <a href="https://scans.gradle.com/s/5dgkmj22kdgls" rel="nofollow noopener">https://scans.gradle.com/s/5dgkmj22kdgls</a></p>
<p>All of this was August 22 2019</p>

**Comment by augur** [comment]
<p>I’ve exactly the same issue. Have you found any solution?</p>

**Comment by clarkbw** [comment, marked as answer]
<p>This should work now</p>
<p>I forked your project and ran <code>gradle publish</code> which works without errors.</p>
<p><a href="https://github.com/clarkbw/github-registry-repro/packages/38910" rel="nofollow noopener">https://github.com/clarkbw/github-registry-repro/packages/38910</a></p>
<p>BTW we’ve added a new help section for Gradle here: <a href="https://help.github.com/en/articles/configuring-gradle-for-use-with-github-package-registry" rel="nofollow noopener">https://help.github.com/en/articles/configuring-gradle-for-use-with-github-package-registry</a></p>

**Comment by JoePotter** [comment]
<p>Hey <a class="mention" href="/clarkbw">@clarkbw</a>,</p>
<p>Still having issues with this on my end. When declaring the implementation for this in an Android project, how are you writing it so it finds the files? I can’t seem to get it to point to the files in that fork of yours correctly for the life of me.</p>
<p>Have tried </p>
<pre><code>implementation 'com.trevjonez:github-registry-repro:1.1'
</code></pre>
<p>and pretty much every other variation I can think of. Have set up the maven url to point to the repository correctly and provided authentication, but its not constructing a valid path to get the files from.</p>
<p>Any help would be greatly appreciated!</p>
<p>Edit:</p>
<p>This is now working, my error was when setting the url in the maven block, we were only including the OWNER in the url, not the OWNER/REPOSITORY (on both the publishing and implementation sides). Though this does lead to published packages having the repository name in the name of the package itself, which isn’t ideal, it fixes it enough to use which is great.</p>

**Comment by breandan** [comment]
<p>Hi <a href="/t5/user/viewprofilepage/user-id/3492">@clarkbw</a>, thanks for posting this. FYI the Kotlin DSL instructions are broken on Gradle 5.6.3. I had to change this line:</p>
<pre><code>register("gpr") {
</code></pre>
<p>to:</p>
<pre><code>register("gpr", MavenPublication::class) {
</code></pre>

**Comment by mgodave** [comment]
<p>This does not appear to be solved. I’m still seeing this issue with my project.</p>

**Comment by trevjonez** [comment]
<p>can you be more specific about your issue?</p>

**Comment by clhols** [comment]
<p>I have tried to publish a package from actions by running “gradle publish”, but is getting:</p>
<pre><code>Could not transfer artifact dk.youtec:appupdater:jar:1.0.1 from/to remote (https://maven.pkg.github.com/clhols/appupdater-plugin): Could not write to resource 'dk/youtec/appupdater/1.0.1/appupdater-1.0.1.jar'

Could not transfer artifact dk.youtec:appupdater:pom:1.0.1 from/to remote (https://maven.pkg.github.com/clhols/appupdater-plugin): Could not write to resource 'dk/youtec/appupdater/1.0.1/appupdater-1.0.1.pom'
&gt; Task :publishAppupdaterPluginPublicationToGitHubPackagesRepository FAILED

7 actionable tasks: 7 executed
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':publishAppupdaterPluginPublicationToGitHubPackagesRepository'.
&gt; Failed to publish publication 'appupdaterPlugin' to repository 'GitHubPackages'
   &gt; Could not deploy to remote repository | Failed to deploy artifacts: Could not transfer artifact dk.youtec:appupdater:jar:1.0.1 from/to remote (https://maven.pkg.github.com/clhols/appupdater-plugin): Could not write to resource 'dk/youtec/appupdater/1.0.1/appupdater-1.0.1.jar'
</code></pre>
<p>build.gradle.kts looks like this:</p>
<pre><code>publishing {
    repositories {
        maven {
            name = "GitHubPackages"
            url = uri("https://maven.pkg.github.com/clhols/appupdater-plugin")
            credentials {
                username = project.findProperty("gpr.user") as String? ?: System.getenv("USERNAME")
                password = project.findProperty("gpr.key") as String? ?: System.getenv("PASSWORD")
            }
        }
    }
    publications {
        register("appupdaterPlugin", MavenPublication::class.java) {
            groupId = "dk.youtec"
            artifactId = "appupdater"
            version = "1.0.1"

            from(components["java"])
        }
    }
}
</code></pre>
<p>Actions build.yml looks like this:</p>
<pre><code>name: CI

on: [push]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v1
    - name: set up JDK 1.8
      uses: actions/setup-java@v1
      with:
        java-version: 1.8
    - name: Build appupdater plugin
      run: |
        ./gradlew publish
</code></pre>

**Comment by clhols** [comment]
<p>I figured it out. I just had to add the properties to the gradle command in build.yml myself, like so:</p>
<pre><code>-name: Build appupdater plugin
run: |
./gradlew build
./gradlew -Pgpr.user=clhols -Pgpr.key=${{ secrets.GITHUB_TOKEN }} publishAppupdaterPluginPublicationToGitHubPackagesRepository
</code></pre>

**Comment by jef** [comment]
<p>I ended up using</p>
<pre><code>register&lt;MavenPublication&gt;("gpr"){
</code></pre>
<p>Basically the same thing, just throwing that out there :slight_smile:</p>
